### PR TITLE
Cleanup resources in Java client

### PR DESCRIPTION
**Pull request by BrennanConroy**
Before, the chat sample would just hang when main exited. This is because Java is waiting for threads to close before exiting the program. Now at least if main exits (without the [try resource](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html)) the program will close after ~1 minute which is how long OkHttp will keep its internal threads open. Otherwise, using the try resource pattern, the program will close immediately.

We'll need to discuss the API additions.

### CODE REVIEW

File: src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java [hunk 1]
```diff
@@ -531,6 +534,17 @@ private void stopConnection(String errorMessage) {
             transportEnum = TransportEnum.ALL;
             this.localHeaders.clear();
             this.streamMap.clear();
+            if (this.pingTimer != null) {
+                this.pingTimer.cancel();
+            }
```
Comment (halter73) [inline src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java:539]: The pingTimer is already being canceled a few lines above in this method. There was a community PR that added it recently.

File: src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java [hunk 2]
```diff
@@ -1097,4 +1111,19 @@ public InvocationRequest tryRemoveInvocation(String id) {
             return handlers.get(0).getClasses();
         }
     }
+
+    @Override
+    public void close() throws Exception {
+        stop().subscribe(() -> {
```
Comment (halter73) [inline src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java:1117]: Shouldn't this block on stop() and cleanup(), so when close() returns the HubConnection is completely closed?

File: src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java [hunk 3]
```diff
@@ -1097,4 +1111,19 @@ public InvocationRequest tryRemoveInvocation(String id) {
...
+            this.cleanup();
+        }, e -> {
+            this.cleanup();
+        });
+    }
+
+    private void cleanup() {
+        if (this.httpClient != null) {
+            this.httpClient.close();
```
Comment (halter73) [inline src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java:1126]: Should we be doing this if we didn't new up the httpClient in the ctor?

File: src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java (hunk 2 above)
Comment (BrennanConroy) [reply src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java:1117]: Yeah, I guess this is like calling Dispose on an object with async close, you need to block on the task.

**Review by jkotalik — APPROVED** [review]

File: src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java [hunk 4]
```diff
@@ -24,13 +24,14 @@
 /**
  * A connection used to invoke hub methods on a SignalR Server.
  */
-public class HubConnection {
+public class HubConnection implements AutoCloseable {
```
Comment (BrennanConroy) [inline src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java:27]: @anurse should we go through API review for this, or do it among the SignalR folks?

Comment (anurse) [reply src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java:27]: It probably should go through API review but it’s pretty minor. Could be done on email. Is this basically Java’s “IDisposable”? If so, I don’t see any concern with adding it.

Comment (BrennanConroy) [reply src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java:27]: > Is this basically Java’s “IDisposable”?

Basically
https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html

**Comment by BrennanConroy** [comment]
API approved over email.

**Review by halter73 — APPROVED** [review]

**Pull request merged**